src/qibo/models/dbi/double_bracket.py

```diff
@@ -156,6 +203,98 @@ def hyperopt_step(
+        e.g. $n=2$: $2\Trace(\sigma(\Gamma_1 + s\Gamma_2 + s^2/2\Gamma_3)\sigma(\Gamma_0 + s\Gamma_1 + s^2/2\Gamma_2))
+        Args:
+            n (int, optional): the order to which the loss function is expanded. Defaults to 4.
+            n_max (int, optional): maximum order allowed for recurring calls of `polynomial_step`. Defaults to 5.
+            d (np.array, optional): diagonal operator, default as $\delta(H)$.
+            backup_scheduling (`DoubleBracketScheduling`): the scheduling method to use in case no real positive roots are found.
+        """
+
+        if d is None:
+            d = self.diagonal_h_matrix
+
+        if backup_scheduling is None:
+            backup_scheduling = DoubleBracketScheduling.grid_search
+
+        def sigma(h: np.array):
+            return h - self.backend.cast(np.diag(np.diag(self.backend.to_numpy(h))))
+
+        def Gamma(k: int):
+            r"""Computes the k_th Gamma function i.e $\Gamma_k=[W,...,[W,[W,H]]...]$, where we take k nested commutators with $W = [D, H]$"""
+            if k == 0:
+                return self.h.matrix
+            else:
+                W = self.commutator(d, sigma(self.h.matrix))
+                result = self.h.matrix
+                for _ in range(k):
+                    result = self.commutator(W, result)
+            return result
+
+        # list starting from s^n highest order to s^0
+        sigma_gamma_list = np.array([sigma(Gamma(k)) for k in range(n + 2)])
+        exp_list = np.array([1 / math.factorial(k) for k in range(n + 1)])
+        # coefficients for rotation with [W,H] and H
+        c1 = [
+            exp_coef * delta_gamma
+            for exp_coef, delta_gamma in zip(exp_list, sigma_gamma_list[1:])
+        ]
+        c2 = [
+            exp_coef * delta_gamma
+            for exp_coef, delta_gamma in zip(exp_list, sigma_gamma_list[:-1])
+        ]
+        # product coefficient
+        trace_coefficients = [0] * (2 * n + 1)
+        for k in range(n + 1):
+            for j in range(n + 1):
+                power = k + j
+                product_matrix = c1[k] @ c2[j]
+                trace_coefficients[power] += 2 * np.trace(product_matrix)
+        roots = np.roots(list(reversed(trace_coefficients[: n + 1])))
+        error = 1e-3
+        real_positive_roots = [
+            np.real(root)
+            for root in roots
+            if np.imag(root) < error and np.real(root) > 0
+        ]
+        # solution exists, return minimum s
+        if len(real_positive_roots) > 0:
+            return min(real_positive_roots)
+        # solution does not exist, resort to backup scheduling
+        elif (
+            backup_scheduling == DoubleBracketScheduling.polynomial_approximation
+            and n < n_max + 1
+        ):
+            return self.polynomial_step(
+                n=n + 1, d=d, backup_scheduling=backup_scheduling
+            )
+        else:
+            return self.choose_step(d=d, scheduling=backup_scheduling)
```

Comment:
I will expect this function to return the result of the polynomial optimization, if it fails it would be better to return a `None`, instead of a result of a different optimization.